src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java

```diff
@@ -399,6 +463,143 @@ public Configuration getHadoopConfiguration() {
+                topic,
+                topixRegexCaptureGroup.pattern(),
+                TOPIC_REGEX_CAPTURE_GROUP_CONFIG
+            )
+        );
+      }
+
+      // make sure that all references to captured groups actually exist
+      if (maxIndexParts > 0 && maxIndexParts > matcher.groupCount()) {
+        throw new ConfigException(
+            String.format(
+                "Topic %s must have at least %d capture groups using regex pattern %s, "
+                    + "but actually had %d capture groups.",
+                topic,
+                maxIndexParts,
+                topixRegexCaptureGroup.pattern(),
+                matcher.groupCount()
+            )
+        );
+      }
+
+      for (int index = 1; index < matcher.groupCount() + 1; index++) {
+        dir = dir.replace("${" + index + "}", matcher.group(index));
+      }
+    }
+
+    return dir;
+  }
+
+  /**
+   * Finds all instances of `${[0-9]+}` in a string and returns the maximum integer.
+   * @return int - largest number found in the regex pattern - 0 if none found
+   */
+  private int getMaxIndexToReplace(String string) {
+
+    List<Integer> toReplace = new ArrayList<>();
+    Pattern pattern = Pattern.compile("\\$\\{(\\d+)}");
+    Matcher partsMatcher = pattern.matcher(string);
+
+    while (partsMatcher.find()) {
+      String part = partsMatcher.group();
+      if (!part.isEmpty() && partsMatcher.groupCount() == 1) {
+        toReplace.add(Integer.valueOf(partsMatcher.group(1))); // should be the first group
+      }
+    }
+
+    return toReplace.isEmpty() ? 0 : Collections.max(toReplace);
+  }
+
+  /**
+   * Validates that the `topics.dir`, `logs.dir`, and `topic.regex.capture.group` configs are all
+   * within valid ranges.
+   */
+  private void validateDirsAndRegex() {
+    if (topicDirGroupsMaxIndex > 0 && topixRegexCaptureGroup == null) {
+      throw new ConfigException(
+          TOPICS_DIR_CONFIG + " cannot contain ${} without a valid "
+              + TOPIC_REGEX_CAPTURE_GROUP_CONFIG + " being configured."
+      );
+    }
+
+    if (logDirGroupsMaxIndex > 0 && topixRegexCaptureGroup == null) {
+      throw new ConfigException(
+          LOGS_DIR_CONFIG + " cannot contain ${} without a valid "
+              + TOPIC_REGEX_CAPTURE_GROUP_CONFIG + " being configured."
+      );
+    }
+
+    validateReplacements(TOPICS_DIR_CONFIG);
+    validateReplacements(LOGS_DIR_CONFIG);
+  }
+
+  /**
+   * Validates that the config has no invalid substitutions
+   *
+   * @param config - String - the config to validate
+   */
+  private void validateReplacements(String config) {
+    // remove all valid ${} substitutions
+    Pattern pattern = Pattern.compile("\\$\\{(\\d+)}");
+    Matcher partsMatcher = pattern.matcher(getString(config));
+    String dir = partsMatcher.replaceAll("").replace("${topic}", "");
+
+    // check for invalid ${} substitutions
+    Pattern invalid = Pattern.compile("\\$\\{.*}");
```

Comment:
nit: these patterns can be made `private static final` and reused.

Reply:
this is only run once at ConfigConstruction essentially being static - but I can make it static